docker/pusher.go

```diff
@@ -23,46 +29,66 @@ func NewPusher() *Pusher {
 }
 
 // Push the image of the Function.
-func (n *Pusher) Push(f bosonFunc.Function) (digest string, err error) {
-	// Check for the docker binary explicitly so that we can return
-	// an extra-friendly error message.
-	_, err = exec.LookPath("docker")
-	if err != nil {
-		err = errors.New("please install 'docker'")
-		return
-	}
+func (n *Pusher) Push(ctx context.Context, f bosonFunc.Function) (digest string, err error) {
 
 	if f.Image == "" {
 		return "", errors.New("Function has no associated image.  Has it been built?")
 	}
 
-	// set up the command, specifying a sanitized project name and connecting
-	// standard output and error.
-	cmd := exec.Command("docker", "push", f.Image)
+	parts := strings.Split(f.Image, "/")
+	if len(parts) < 1 {
+		return "", errors.Errorf("invalid image name: %q", f.Image)
+	}
+	registry := parts[0]
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create docker api client")
+	}
+
+	credentials, err := config.GetAllCredentials(nil)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get credentials")
+	}
+
+	var opts types.ImagePushOptions
+
+	c := credentials[registry]
+	opts.RegistryAuth = base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"username":"%s","password":"%s"}`, c.Username, c.Password)))
```

Comment:
Hmm... my docker config looks like this:

```
{
  "HttpHeaders": {
    "User-Agent": "Docker-Client/19.03.13 (linux)"
  },
  "auths": {
    "https://index.docker.io/v1/": {
      "auth": "REDACTED"
    },
    "quay.io": {
      "auth": "REDACTED"
    },
    "registry.svc.ci.openshift.org": {
      "auth": "REDACTED"
    }
  }
}
```
Will this work? Maybe I'm making too many assumptions about what the docker API is doing...

Reply:
Fixed: if the `namespace/image:tag` format is used then `docker.io` registry is assumed.

Reply:
It should. Please try it out.